test/unit/upload.js

```diff
@@ -36,66 +41,218 @@ describe('Upload Methods', function () {
-          sinon.assert.calledOnce(mockInstance.newContainer)
-          sinon.assert.calledWithExactly(
-            mockInstance.newContainer,
-            'mockContainerId'
-          )
-        })
+  describe('_recusivelyCreateDirectories', function () {
+    var parentPath = '/cwd'
+
+    beforeEach(function () {
+      sinon.stub(upload, '_createDirectory')
+    })
+
+    afterEach(function () {
+      upload._createDirectory.restore()
     })
 
-    it('should upload the file', function () {
-      return assert.isFulfilled(upload.uploadFile(mockArgs))
+    describe('with non-existing directories', function () {
+      beforeEach(function () {
+        upload._createDirectory.resolves()
+      })
+
+      it('should not create any directories if none are passed', function () {
+        return assert.isFulfilled(upload._recusivelyCreateDirectories(mockContainer, parentPath, '/'))
+          .then(function () {
+            assert.equal(upload._createDirectory.callCount, 0)
+          })
+      })
+
+      it('should create all directories', function () {
+        return assert.isFulfilled(upload._recusivelyCreateDirectories(mockContainer, parentPath, '/0/1/2/3/4/'))
+          .then(function () {
+            assert.equal(upload._createDirectory.callCount, 5)
+            var arr = [0, 1, 2, 3, 4]
+            arr.forEach(function (num) {
+              var call = upload._createDirectory.getCall(num)
+              call.calledWith(sinon.match.any, parentPath, num)
+            })
+          })
+      })
+
+      it('should not care whether theres a forward slash at the end', function () {
+        return assert.isFulfilled(upload._recusivelyCreateDirectories(mockContainer, parentPath, '0/1/2'))
+          .then(function () {
+            assert.equal(upload._createDirectory.callCount, 3)
+            var arr = [0, 1, 2]
+            arr.forEach(function (num) {
+              var call = upload._createDirectory.getCall(num)
+              call.calledWith(sinon.match.any, parentPath, num)
+            })
+          })
+      })
+    })
+
+    describe('with existing directories', function () {
+      beforeEach(function () {
+        upload._createDirectory.rejects()
+      })
+
+      it('should create all directories', function () {
+        return assert.isFulfilled(upload._recusivelyCreateDirectories(mockContainer, parentPath, '/0/1/2/3/4/'))
+          .then(function () {
+            assert.equal(upload._createDirectory.callCount, 5)
+            var arr = [0, 1, 2, 3, 4]
+            arr.forEach(function (num) {
+              var call = upload._createDirectory.getCall(num)
+              call.calledWith(sinon.match.any, parentPath, num)
+            })
+          })
+      })
+    })
+  })
+
+  describe('_createFile', function () {
+    var path = '/cwd/hello/world'
+    var file = 'hello.js'
+
+    it('should call the container.createFile method', function () {
+      upload._createFile(mockContainer, path, file, mockFileData)
```

Comment:
this one is breaking the pattern. since it returns a promise, you should (at minimum) do your assertion below in a `.then`, and return the promise chain